Fix typescript errors and add a new config file for compilation checks

ankita240796 opened the pull request:
## Explanation
Fixes typescript errors introduced in the codebase and adds a new config file for compilation checks.

The import of files gives an error: `error TS2691: An import path cannot end with a '.ts' extension.` Since we are planning to remove the import blocks, I have added an ts-ignore comment there to ignore these errors. 

There are some import statements which are not in TODO statement to be removed later. Is that intentional or TODO statement was missed there? /cc @YashJipkate 

I have added a new config file which checks for errors in the complete code but I don't think it is exiting the compilation if errors are found. I was not able to find any specific way to show that the compilation failed due to errors. @vojtechjelinek  do you have any suggestions for the same?

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

ankita240796 committed:
Fix errors

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 committed:
Remove extra conversion

ankita240796 committed:
Fix lint

ankita240796 committed:
Disable frontend test

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 committed:
Enable test

YashJipkate commented on core/templates/dev/head/domain/exploration/AnswerStatsObjectFactory.ts:
```diff
@@ -17,7 +17,7 @@
  */
 
-import * as cloneDeep from 'lodash/cloneDeep';
+import cloneDeep from 'lodash/cloneDeep';
```
Please make sure to test this. This can become tricky at times.

ankita240796 replied:
Sure, will do. What issues did you face here (It only interfered with Karma tests or with the dev server as well?)

seanlip replied:
@YashJipkate could you please reply to this?

YashJipkate replied:
So sorry, I somehow missed this.

> What issues did you face here (It only interfered with Karma tests or with the dev server as well?)

Yes, those were only karma-specific, worked fine on dev server.

ankita240796 replied:
Ok, I checked by creating and publishing a simple exploration on dev server and it worked fine. Karma tests passed too!

YashJipkate reviewed (approved):
LGTM Thanks @ankita240796. 
About the TODOs, I guess you are talking about the ones in the spec files. The imports inside the blocks are to be removed on upgrade but the others are required. Hope that answers your question. 
Please let me know if you need any more explanation.
Thanks!

ankita240796 commented:
Hi @YashJipkate, using `.ts` in import statements gives a typescript error as follows:

```
core/templates/dev/head/AppSpec.ts(21,35): error TS2691: An import path cannot end with a 
'.ts' extension. Consider importing 'domain/exploration/RuleObjectFactory' instead.
```

I just added `ts-ignore` to the blocks which are to be removed as I did not find any other way to change the imports. Since, the imports in spec files are not be removed, I think I will need to figure out some way to change the imports there rather than using `ts-ignore`. Do you have any suggestions for the same? It is totally fine if you are busy with your GSoC project and do not have time to look into this. Thanks!

Commit:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

vojtechjelinek commented on tsconfig.json:
```diff
@@ -9,7 +9,8 @@
     "experimentalDecorators": true,
-    "emitDecoratorMetadata": true
+    "emitDecoratorMetadata": true,
+    "esModuleInterop": true
```
Why is this added?

ankita240796 replied:
Hi @vojtechjelinek, the flag allows default imports from modules with no default export.  Importing from `lodash/cloneDeep` using following syntax: `import * as cloneDeep from 'lodash/cloneDeep';` results in a compilation error as follows: `This module can only be referenced with ECMAScript imports/exports by turning on the 'esModuleInterop' flag and referencing its default export.` So, I added this flag and switched to default import instead of using `* _ as` syntax.

Further, I tried to include the files in tsconfig and then use import statements without `.ts` extension but that results in an error of module not found.

vojtechjelinek reviewed (changes requested):
> I have added a new config file which checks for errors in the complete code but I don't think it is exiting the compilation if errors are found. I was not able to find any specific way to show that the compilation failed due to errors.

Do you mean webpack compilation here?

Regarding the `.ts` extensions: for some reason we couldn't import files without the extension, I think it might be related to the fact the files were not included in the *tsconfig.json*, could you please try to add the include the files in *tsconfig.json* and then try to use the import statements without `.ts` extension?

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 committed:
Add default imports

vojtechjelinek commented:
@ankita240796 Could you please try one more thing? As per this https://stackoverflow.com/questions/43595555/webpack-cant-resolve-typescript-modules adding `extensions` to `resolve` should fix this. You might need to add all the extensions that we pass through webpack like `.html` or `.svg`.

ankita240796 commented:
Hi @vojtechjelinek, I tried adding resolve as suggested by you but it does not work either. The error is a typescript compile time error:

If `.ts` is used in imports, the error says:

```
ERROR in /Users/saxenaankita/Desktop/opensrc/oppia/core/templates/dev/head/AppSpec.ts
ERROR in /Users/saxenaankita/Desktop/opensrc/oppia/core/templates/dev/head/AppSpec.ts(22,3):
TS2691: An import path cannot end with a '.ts' extension. Consider importing 'domain/exploration/ParamChangeObjectFactory' instead.
```

If we import without the extensions, the error is as follows:

```
ERROR in /Users/saxenaankita/Desktop/opensrc/oppia/core/templates/dev/head/AppSpec.ts
ERROR in /Users/saxenaankita/Desktop/opensrc/oppia/core/templates/dev/head/AppSpec.ts(23,35):
TS2307: Cannot find module 'domain/exploration/RuleObjectFactory'.
```
 
Sorry, I missed one of your previous questions:

> I have added a new config file which checks for errors in the complete code but I don't think it is exiting the compilation if errors are found. I was not able to find any specific way to show that the compilation failed due to errors.

> Do you mean webpack compilation here? 

Yes, I mean webpack compilation.

vojtechjelinek commented:
> I have added a new config file which checks for errors in the complete code but I don't think it is exiting the compilation if errors are found. I was not able to find any specific way to show that the compilation failed due to errors.

I'm not sure why the webpack is not failing when the type errors exist, you might need to look into the `ForkTsCheckerWebpackPlugin` docs.

Regarding the issue with `.ts` extension, my last idea is to combine both previously mentioned approaches: 

- add `extensions` to `resolve`
- add files into `tsconfig.json`

If that doesn't work the I think it is okay to keep the `@ts-ignore` and filing an issue about this.

ankita240796 commented:
Hi @vojtechjelinek, I am checking for the import issue and I will update you on that. 

I also filed an issue for webpack compilation not failing in case of errors, and [here](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/issues/324#issuecomment-519034281) is the solution I got.

I just wanted to know should we use it in dev mode also, since we need to alert the user if there are errors locally as well so that the user can fix them before issuing a PR.

What are your suggestions for the same?

ankita240796 committed:
Enable webpack to halt in case of errors

vojtechjelinek commented:
I don't think we should do that in dev mode. Should we maybe do the typescript compilation as part of the lint checks?

ankita240796 commented:
That will slow down the linting process, so I don't think we should compile during lint.

I think it is fine to halt the webpack compilation in only prod mode and update the wiki to have instructions to check errors after starting a dev server. Contributors are expected to run a dev server and test various features before submitting a PR and so, it should be ok to check errors and fix them too in the same process.

What do you think?

EDIT: There is one other thing which we can do - add `async: false` to karma tests as well so that the frontend tests fail if there is compilation error.

ankita240796 committed:
Add async for only prod mode

ankita240796 committed:
Fix import issues

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 committed:
Fix subsequent var dec issue

ankita240796 committed:
Add semicolon

ankita240796 committed:
Fix test

ankita240796 commented:
Hi @vojtechjelinek, I was finally able to resolve both the issues:

- I have removed `.ts` from imports as well as `@ts-ignore` comments. Adding `extensions` to resolve in webpack config and `baseUrl` and `paths` in tsconfig fixed the issue.
- For halting webpack compilation, I have added `async: false` in both karma webpack config and prod webpack config. Is that ok?

PTAL, Thanks!

ankita240796 committed:
Test that webpack fails on compilation errors

ankita240796 committed:
Webpack fails, revert test commit

ankita240796 commented:
Hi @seanlip, can we get this in quickly with only approvals from @vojtechjelinek and @YashJipkate to prevent introduction of more errors? Thanks!

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 committed:
Remove new errors

seanlip commented on core/templates/dev/head/services/CsrfTokenService.ts:
```diff
@@ -18,12 +18,8 @@
 // This needs to be imported first instead of using the global definition
 // because Angular doesn't support global definitions and every library used
-// needs to be imported explicitly. Also, default imports do not go do not go
-// down well with Karma and thus the import-as syntax.
```
@YashJipkate -- just to confirm, was the old stuff here (the part that's now been deleted) incorrect? Would be good to get an understanding of where the discrepancy arose. /cc @kevinlee12

ankita240796 replied:
I think it was not wrong but we missed the flag `esModuleInterop` in tsconfig which allows for default imports. After adding the flag, imports work fine.

YashJipkate replied:
@seanlip I have inferred this myself since the default export stuff wasn't working (maybe because of some Webpack or TS config state). 
Thanks, @ankita240796 for giving a fix to this!

seanlip reviewed (approved):
Hi @ankita240796 -- once you get responses from @YashJipkate on the two open comments I left, and an approval from @vojtechjelinek, then yes this can be merged (just ping me). Thanks!

ankita240796 commented:
Hi @seanlip, Sure I will ping you once both the comments are resolved and I get approval from @vojtechjelinek, Thanks!

ankita240796 committed:
Merge branch 'fix-ts' of github.com:ankita240796/oppia into fix-ts

vojtechjelinek reviewed (approved):
Good job! I think adding `async: false` is okay for now, we will see how that works in the dev workflow.

ankita240796 commented:
Thanks @vojtechjelinek!

seanlip commented:
Confirming this can be merged -- please let me know once tests pass and I'll do it. Thanks!

ankita240796 commented:
Sure, thanks @seanlip!

kevinlee12 commented:
@seanlip, all tests passed!

seanlip commented:
Thanks, merging!
